packages/cli-tools/src/resolveTransitiveDeps.ts

```diff
@@ -0,0 +1,421 @@
+  availableVersions: string[],
+): string | null {
+  const sortedVersions = availableVersions
+    .filter((version) =>
+      ranges.every((range) => semver.satisfies(version, range)),
+    )
+    .sort(semver.rcompare);
+
+  return sortedVersions.length > 0 ? sortedVersions[0] : null;
+}
+
+export function findDependencyPath(
+  dependencyName: string,
+  rootPath: string,
+  parentPath: string,
+) {
+  let dependencyPath;
+  const topLevelPath = path.join(rootPath, 'node_modules', dependencyName);
+  const nestedPath = path.join(parentPath, 'node_modules', dependencyName);
+
+  if (fs.existsSync(topLevelPath)) {
+    dependencyPath = topLevelPath;
+  } else if (fs.existsSync(nestedPath)) {
+    dependencyPath = nestedPath;
+  }
+
+  return dependencyPath;
+}
+
+export function collectDependencies(root: string): Map<string, DependencyData> {
+  const dependencies = new Map<string, DependencyData>();
+
+  const checkDependency = (dependencyPath: string) => {
+    const packageJsonPath = path.join(dependencyPath, 'package.json');
+    const packageJson = require(packageJsonPath);
+
+    if (dependencies.has(packageJson.name)) {
+      const dependency = dependencies.get(packageJson.name) as DependencyData;
+
+      if (
+        dependencyPath !== dependency.path &&
+        dependency.duplicates?.every(
+          (duplicate) => duplicate.path !== dependencyPath,
+        )
+      ) {
+        dependencies.set(packageJson.name, {
+          ...dependency,
+          duplicates: [
+            ...dependency.duplicates,
+            {
+              path: dependencyPath,
+              version: packageJson.version,
+              peerDependencies: packageJson.peerDependencies,
+            },
+          ],
+        });
+      }
+      return;
+    }
+
+    dependencies.set(packageJson.name, {
+      path: dependencyPath,
+      version: packageJson.version,
+      peerDependencies: packageJson.peerDependencies,
+      duplicates: [],
+    });
+
+    for (const dependency in {
+      ...packageJson.dependencies,
+      ...(root === dependencyPath ? packageJson.devDependencies : {}),
+    }) {
+      const depPath = findDependencyPath(dependency, root, dependencyPath);
+
+      if (depPath) {
+        checkDependency(depPath);
+      }
+    }
+  };
+
+  checkDependency(root);
+
+  return dependencies;
+}
+
+export function filterNativeDependencies(
+  root: string,
+  dependencies: Map<string, DependencyData>,
+) {
+  const depsWithNativePeers = new Map<string, Map<string, string>>();
+
+  dependencies.forEach((value, key) => {
+    if (value.peerDependencies) {
+      const nativeDependencies = new Map<string, string>();
+
+      Object.entries(value.peerDependencies).forEach(([name, versions]) => {
+        const dependencyPath = findDependencyPath(name, root, value.path);
+
+        if (dependencyPath) {
+          const iosPath = path.join(dependencyPath, 'ios');
+          const androidPath = path.join(dependencyPath, 'android');
```

Comment:
This is introducing new logic for determining whether this is a native dependency or not, and project can configure their own through `react-native.config.js`. Instead we should reuse existing mechanism so it's deterministic. Check how config calls platforms `projectConfig`: 
https://github.com/react-native-community/cli/blob/d6951445f967798de78c041dd2d9cc3f1b21e53a/packages/cli-config/src/loadConfig.ts#L43-L46 and 
Here's a reference implementation for Android:
https://github.com/react-native-community/cli/blob/d6951445f967798de78c041dd2d9cc3f1b21e53a/packages/cli-platform-android/src/config/index.ts#L34-L86